x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -727,16 +814,313 @@ public void runAsDenied(String requestId, Authentication authentication, RestReq
+                    }
+                    // the toXContent methods of the {@code RoleMapperExpression} instances do a good job
+                    builder.field("rule", putRoleMappingRequest.getRules())
+                    .field("enabled", putRoleMappingRequest.isEnabled());
+                    if (putRoleMappingRequest.getMetadata() != null && false == putRoleMappingRequest.getMetadata().isEmpty()) {
+                        builder.field("metadata", putRoleMappingRequest.getMetadata());
+                    }
+                    builder.endObject() // role_mapping
+                    .endObject();
+            logEntry.with(PUT_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(SetEnabledRequest setEnabledRequest) throws IOException {
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    .startObject("user")
+                    .field("name", setEnabledRequest.username())
+                    .endObject() // user
+                    .endObject();
+            if (setEnabledRequest.enabled() != null && setEnabledRequest.enabled()) {
+                logEntry.with(EVENT_ACTION_FIELD_NAME, "enable_user");
+                logEntry.with(ENABLE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            } else {
+                logEntry.with(EVENT_ACTION_FIELD_NAME, "disable_user");
+                logEntry.with(DISABLE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            }
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(PutPrivilegesRequest putPrivilegesRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "put_privileges");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    // toXContent of {@code ApplicationPrivilegeDescriptor} does a good job
+                    .field("privileges", putPrivilegesRequest.getPrivileges())
+                    .endObject();
+            logEntry.with(PUT_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(CreateApiKeyRequest createApiKeyRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "create_apikey");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject();
+            withRequestBody(builder, createApiKeyRequest);
+            builder.endObject();
+            logEntry.with(CREATE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(GrantApiKeyRequest grantApiKeyRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "create_apikey");
```

Comment:
Should this be `grant_api_key` and save the nested `grant` field?

Reply:
If you ask me, I think it makes sense to have both the ways to create api keys nested under `"create": {"apikey":{...` with different internal fields.

Reply:
I see you are trying to hold a connection between the value of `event.action` and the value of the top level payload key. I think this consistency has value, so let's keep it as is.